ajaxterm.1
.TH ajaxterm "1" "Jul 2006" "ajaxterm 0.7" "User commands"
.SH NAME
ajaxterm \- Web based terminal written in python

.SH DESCRITPION
\fBajaxterm\fR is a web based terminal written in python and some AJAX
javascript for client side.
It can use almost any web browser and even works through firewalls.

.SH USAGE
\fBajaxterm\fR [options]

.SH OPTIONS
A summary of the options supported by \fBajaxterm\fR is included below.
.TP
\fB\-h\fR, \fB\-\-help\fR
show the help message and exit
.TP
\fB\-p\fR PORT, \fB\-\-port\fR=\fIPORT\fR
Set the TCP port (default: 8022)
.TP
\fB\-c\fR CMD, \fB\-\-command\fR=\fICMD\fR
set the command (default: /bin/login or ssh localhost)
.TP
\fB\-l\fR, \fB\-\-log\fR
log requests to stderr (default: quiet mode)
.TP
\fB\-d\fR, \fB\-\-daemon\fR
run as daemon in the background
.TP
\fB\-P\fR PIDFILE, \fB\-\-pidfile\fR=\fIPIDFILE\fR
set the pidfile (default: /var/run/ajaxterm.pid)
.TP
\fB\-i\fR INDEX_FILE, \fB\-\-index\fR=\fIINDEX_FILE\fR
default index file (default: ajaxterm.html)
.TP
\fB\-u\fR UID, \fB\-\-uid\fR=\fIUID\fR
Set the daemon's user id
.TP
\fB\-s\fR SERVERPORT, \fB\-\-serverport\fR=\fISERVERPORT\fR
Use a different port than 22 to connect to the ssh
server
.TP
\fB\-t\fR TOKEN, \fB\-\-token\fR=\fITOKEN\fR
Set authorization token. If this option is passed on the command line, the
given token must be used when accessing ajaxterm (by adding a ?token=something
request argument).
.TP
\fB\-T\fR SECONDS, \fB\-\-terminate-on-idle\fR=\fISECONDS\fR
Terminate if idle for more than SEC seconds (0 means never)

.SH AUTHOR
Antony Lesuisse <al@udev.org>

Adopted to UTF-8 by Sergej Pupykin <ps@lx-ltd.ru>

This manual page was written for the Debian system by
Julien Valroff <julien@debian.org> (but may be used by others).

.SH "REPORTING BUGS"
Report any bugs to the author: Antony Lesuisse <al@udev.org>

.SH COPYRIGHT
Copyright Antony Lesuisse <al@udev.org>

.SH SEE ALSO
- \fBajaxterm\fR wiki page: http://antony.lesuisse.org/qweb/trac/wiki/AjaxTerm
.br
- \fBajaxterm\fR forum: http://antony.lesuisse.org/qweb/forum/viewforum.php?id=2


ajaxterm.py
#!/usr/bin/env python3
# Copyright 2015 Evgeny Golyshev
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

import fcntl
import optparse
import os
import pty
import select
import signal
import socket  # only for gethostname()
import struct
import sys
import termios
import threading

import tornado.httpserver
import tornado.ioloop
import tornado.web
import tornado.websocket

import terminal

os.chdir(os.path.normpath(os.path.dirname(__file__)))


class IndexHandler(tornado.web.RequestHandler):
    def get(self):
        self.render("index.htm")


class BasicHandler(tornado.web.RequestHandler):
    def get(self):
        self.render("basic.htm")


class ControlPanelHandler(tornado.web.RequestHandler):
    def get(self):
        self.render("control-panel.htm")


class TermSocketHandler(tornado.websocket.WebSocketHandler):
    # clients представляет собой разреженный массив. Словарь
    # отлично подходит для решения этой задачи, но, возможно, в
    # недалеком будущем стоит задуматься об использовании менее
    # универсальной структуры данных.
    clients = {}
    epoll = select.epoll()
    fd = None

    def create(self, width = 80, height = 24):
        pid, fd = pty.fork()
        if pid == 0:
            if os.getuid() == 0:
                cmd = ['/bin/login']
            else:
                # В Python 3.x, в отличии от Python 2.x, для того чтобы строка
                # попала в Терминал, необходимо завершить ее символом \n,
                # который в данном случае эмулирует нажатие клавиши Enter.
                # Однако, необходимость вводить таким образом логин при
                # подключении к машине через ssh является всего лишь временной
                # мерой. В недалеком будущем Терминал будет расширен опцией
                # командной строки, которая позволит указать как логин, так и
                # хост. К примеру, --ssh eugulixes@192.168.0.100.
                sys.stdout.write(socket.gethostname() + ' login: \n')
                login = sys.stdin.readline().strip()

                cmd = ['ssh']
                cmd += ['-oPreferredAuthentications=keyboard-interactive,password']
                cmd += ['-oNoHostAuthenticationForLocalhost=yes']
                cmd += ['-oLogLevel=FATAL']
                cmd += ['-F/dev/null']
                cmd += ['-l', login, 'localhost']

            env = {}
            env["COLUMNS"] = str(width)
            env["LINES"] = str(height)
            env["TERM"] = "linux"
            env["PATH"] = os.environ["PATH"]
            os.execvpe(cmd[0], cmd, env)
        else:
            fcntl.fcntl(fd, fcntl.F_SETFL, os.O_NONBLOCK)
            fcntl.ioctl(fd, termios.TIOCSWINSZ, struct.pack("HHHH", height, width, 0, 0))
            TermSocketHandler.clients[fd] = {
                "client": self,
                "pid": pid,
                "terminal": terminal.Terminal(width, height)
            }
            return fd

    def open(self):
        self.fd = self.create()
        self.epoll.register(self.fd, select.EPOLLIN)

    def on_message(self, request):
        label, data = request.split(',')

        if label == 'key':
            try:
                os.write(self.fd, data.encode("utf8"))
            except (IOError, OSError):
                self.kill(self.fd)
        elif label == 'rsz':
            row, col = data.split('x')
            fcntl.ioctl(self.fd,
                        termios.TIOCSWINSZ,
                        struct.pack("HHHH", int(row), int(col), 0, 0))
            self.clients[self.fd]["terminal"].set_resolution(int(row), int(col))

    def on_close(self):
        self.epoll.unregister(self.fd)
        self.kill(self.fd)

    def kill(self, fd):
        try:
            os.close(fd)
            os.kill(self.clients[fd]["pid"], signal.SIGHUP)
        except (IOError, OSError):
            pass
        del self.clients[fd]

    @classmethod
    def dump(cls, fd):
        buf = os.read(fd, 65536)
        cls.clients[fd]["terminal"].write(buf)
        return cls.clients[fd]["terminal"].dumphtml()


class Application(tornado.web.Application):
    def __init__(self):
        handlers = [
            (r"/", IndexHandler),
            (r"/basic", BasicHandler),
            (r"/control-panel", ControlPanelHandler),
            (r"/termsocket", TermSocketHandler),
        ]
        settings = dict(
            template_path=os.path.join(os.path.dirname(__file__), "templates"),
            static_path=os.path.join(os.path.dirname(__file__), "static"),
        )
        tornado.web.Application.__init__(self, handlers, **settings)


def loop():
    while True:
        events = TermSocketHandler.epoll.poll(1)
        for fd, event in events:
            if event & select.EPOLLIN:
                dump = TermSocketHandler.dump(fd)
                TermSocketHandler.clients[fd]["client"].write_message(dump)


def main():
    usage = "usage: %prog [options]"
    parser = optparse.OptionParser(usage=usage)
    parser.add_option("-p", "--port", default=8888,
                      help="the port on which the web server is listening")

    (options, args) = parser.parse_args()

    thread = threading.Thread(target=loop)
    thread.start()
    http_server = tornado.httpserver.HTTPServer(Application())
    http_server.listen(options.port)
    tornado.ioloop.IOLoop.instance().start()

if __name__ == "__main__":
    main()
